core/src/providers/openai.rs

```diff
@@ -1200,41 +1200,86 @@ impl Embedder for OpenAIEmbedder {
             ));
         }
 
-        // Give priority to `CORE_DATA_SOURCES_OPENAI_API_KEY` env variable
-        match std::env::var("CORE_DATA_SOURCES_OPENAI_API_KEY") {
-            Ok(key) => {
-                self.api_key = Some(key);
-            }
-            Err(_) => match credentials.get("OPENAI_API_KEY") {
+        let use_eu_endpoint: bool = match credentials.get("OPENAI_USE_EU_ENDPOINT") {
+            Some(use_eu_endpoint_str) => use_eu_endpoint_str == "true",
+            None => false,
+        };
+
+        let open_ai_api_key: Option<String> = match credentials.get("OPENAI_API_KEY") {
+            Some(api_key) => Some(api_key.clone()),
+            None => match tokio::task::spawn_blocking(|| std::env::var("OPENAI_API_KEY")).await? {
+                Ok(key) => Some(key),
+                Err(_) => None,
+            },
+        };
+
+        // This is only for logging purposes.
+        let mut key_source = "";
+
+        match use_eu_endpoint {
+            true => match open_ai_api_key {
+                // If use_eu_endpoint is set, we ignore CORE_DATA_SOURCES_OPENAI_API_KEY
                 Some(api_key) => {
                     self.api_key = Some(api_key.clone());
+                    self.host = Some("eu.api.openai.com".to_string());
+                    key_source = "use_eu_endpoint";
                 }
-                None => match std::env::var("OPENAI_API_KEY") {
+                None => Err(anyhow!(
+                    "Credentials or environment variable `OPENAI_API_KEY` is not set."
+                ))?,
+            },
+            false => {
+                // Give priority to `CORE_DATA_SOURCES_OPENAI_API_KEY` env variable
+                match std::env::var("CORE_DATA_SOURCES_OPENAI_API_KEY") {
                     Ok(key) => {
                         self.api_key = Some(key);
+                        key_source = "CORE_DATA_SOURCES_OPENAI_API_KEY";
+
+                        // This block is only for logging purposes, and can be removed when we stop using CORE_DATA_SOURCES_OPENAI_API_KEY
+                        match open_ai_api_key.as_ref() {
+                            Some(front_key) => {
+                                if let Some(core_key) = self.api_key.as_ref() {
+                                    if front_key != core_key {
+                                        fn last_two(s: &str) -> &str {
+                                            // Being safe in case it's an empty string
+                                            if s.len() <= 2 {
+                                                s
+                                            } else {
+                                                &s[s.len() - 2..]
+                                            }
+                                        }
+                                        warn!(
+                                            front_key = last_two(front_key.as_str()),
```

Comment:
Logging the last 2 chars is the only change from the previous change. Just in case we hit this warning, this will be useful.